### PR TITLE
feat: upgrade react router to v6

### PR DESCRIPTION
**Pull request by Syed-Ali-Abbas-Zaidi**
## Ticket
[React Router Upgrade to v6](https://github.com/openedx/platform-roadmap/issues/276).

## Description
This PR upgrades React Router from `v5` to `v6`.

### CODE REVIEW

**Comment by awais-ansari** [comment]
Hi @Syed-Ali-Abbas-Zaidi, Just to confirm. Are you still working on this PR? We are planning the board cleanup. Should I close this PR?

**Comment by Syed-Ali-Abbas-Zaidi** [comment]
> Hi @Syed-Ali-Abbas-Zaidi, Just to confirm. Are you still working on this PR? We are planning the board cleanup. Should I close this PR?

Hi, This PR is currently blocked due to this [issue in react-router v6](https://github.com/remix-run/react-router/pull/10768). This issue was resolved last week but it will be released in a week or two and after that, we will start working on this PR.

**Comment by awais-ansari** [comment]
@Syed-Ali-Abbas-Zaidi We are doing some optimization work and decomposition of tightly coupled components to prevent unwanted re-rendering. This PR will be affected by that work. We need to add a little bit of delay for this upgrade to prevent rework. what is your thought on this? 
Will you continue working on this branch or create a new one after optimization work?

**Comment by awais-ansari** [comment]
@Syed-Ali-Abbas-Zaidi You can start working on this PR.

**Review by sundasnoreen12 — CHANGES_REQUESTED** [review]
These routes display a blank content area in cases where the topic or category does not have any units.

http://localhost:2002/course-v1:edX+DemoX+Demo_Course/topics/i4x-edx-eiorguegnru-course-foobarbaz/posts/

http://localhost:2002/course-v1:edX+DemoX+Demo_Course/category/block-v1:edX+DemoX+Demo_Course+type@sequential+block@edx_introduction/posts

Screenshot 2023-11-28 at 10 23 15 PM


These routes display a blank post content area when topic or category have units.
http://localhost:2002/course-v1:edX+DemoX+Demo_Course/topics/i4x-edx-eiorguegnru-course-foobarbaz/posts/

http://localhost:2002/course-v1:edX+DemoX+Demo_Course/category/block-v1:edX+DemoX+Demo_Course+type@sequential+block@19a30717eff543078a5d94ae9d6c18a5/posts

Screenshot 2023-11-28 at 10 21 17 PM

File: src/discussions/learners/LearnersView.jsx [hunk 1]
```diff
@@ -84,12 +85,9 @@ const LearnersView = () => {
...
-        <Redirect
-          to={{
-            ...location,
-            pathname: Routes.DISCUSSIONS.PATH,
-          }}
-        />
+          <Navigate
+            to={`${truncatePath(location.pathname)}`}
+          />
         )}
```
[the remark this answers is not in the document]
Comment (Syed-Ali-Abbas-Zaidi) [reply src/discussions/learners/LearnersView.jsx:91]: This is the existing code snippet I have just replaced `Redirect` with `Navigate`. But you are right that this component will not be accessed if the `learners flag` is disabled. So I guess we can remove this code snippet from this file, As the redirection is already being handled in the sidebar [here](https://github.com/openedx/frontend-app-discussions/blob/master/src/discussions/discussions-home/DiscussionSidebar.jsx#L97).

File: src/discussions/posts/post/PostLink.jsx [hunk 2]
```diff
@@ -25,6 +25,7 @@ const PostLink = ({
   showDivider,
 }) => {
   const intl = useIntl();
+  const { search } = useLocation();
```
[the remark this answers is not in the document]
Comment (Syed-Ali-Abbas-Zaidi) [reply src/discussions/posts/post/PostLink.jsx:28]: There was some refactoring in the `Link` component of the react-router. Due to this, it does not handle URL query parameters on its own, which is why I explicitly destructured the URL param and passed it to the link component.

File: src/discussions/topics/topic-group/TopicGroupBase.jsx [hunk 3]
```diff
@@ -20,10 +20,15 @@ const TopicGroupBase = ({
   topicsIds,
 }) => {
   const intl = useIntl();
-  const { courseId } = useContext(DiscussionContext);
+  const { search } = useLocation();
```
[the remark this answers is not in the document]
Comment (Syed-Ali-Abbas-Zaidi) [reply src/discussions/topics/topic-group/TopicGroupBase.jsx:23]: There was some refactoring in the `Link` component of the react-router. Due to this, it does not handle URL query parameters on its own, which is why I explicitly destructured the URL param and passed it to the link component.

File: src/data/constants.js [hunk 4]
```diff
@@ -172,7 +171,7 @@ export const Routes = {
       `${BASE_PATH}/my-posts/:postId`,
       `${BASE_PATH}/learners/:learnerUsername/posts/:postId`,
     ],
-    PAGE: `${BASE_PATH}/:page`,
+    PAGE: `${BASE_PATH}/:page/*`,
     PAGES: {
       category: `${BASE_PATH}/category/:category/posts/:postId`,
       topics: `${BASE_PATH}/topics/:topicId/posts/:postId`,
```
[the remark this answers is not in the document]
Comment (sundasnoreen12) [reply src/data/constants.js]: above comment is for line number 186: ${BASE_PATH}/category/:category,
issue still exists

**Comment by Syed-Ali-Abbas-Zaidi** [comment]
> The [route](http://localhost:2002/course-v1:edX+DemoX+Demo_Course/category) is crashing; it should redirect to All Posts for both legacy and new edX.

Fixed. 🤞

**Review by sundasnoreen12 — APPROVED** [review]
PR LGTM

**Review by awais-ansari — CHANGES_REQUESTED** [review]
Some routes are breaking and unexpected behavior. I have listed down all the router bugs and expected behavior in [this doc](https://docs.google.com/spreadsheets/d/1uGR4_zS9bZjlprZt_g-80Mqssd_GZu8Iok3AoAp_XTM/edit?usp=sharing).

**Review by awais-ansari — APPROVED** [review]

**Pull request merged**